codelabs/custom_renderer.md

```diff
@@ -0,0 +1,409 @@
+categories: codelab,blockly,rendering,customization
+status: Draft
+Feedback Link: https://github.com/google/blockly-samples/issues/new
+
+# Custom renderers
+
+## Codelab overview
+
+### What you'll learn
+In this codelab you will learn:
+1. How to define and register a custom renderer.
+1. How to override renderer constants.
+1. How to change the shape of connection notches.
+1. How to set a connection's shape based on its type checks.
+
+### What you'll build
+
+Over the course of this codelab you will build and use four renderers.
+
+1. A minimal custom renderer
+![](assets/custom-renderer/custom_renderer.png)
+1. A renderer with custom constants
+![](assets/custom-renderer/custom_constants.png)
+1. A renderer with custom notches
+![](assets/custom-renderer/custom_notches.png)
+1. A renderer with connection shapes that depend on the type checks of each connection.
+![](assets/custom-renderer/typed_connection_shapes.png)
+### What you'll need
+This codelab assumes that you are already comfortable with using the Blockly playground locally.
+
+## Setup
+
+In this codelab you will add code to the Blockly playground to create and use a new renderer.
+
+When the instructions say to "create and include a new file", you should place that file in the same folder as the playground and include it with a script tag.
+
+```
+<script src="custom_renderer.js"></script>
+```
+
+Note: you must include your custom code *after* including the Blockly library.
+
+## Minimal custom renderer
+
+This is the minimum code needed to define a new renderer.
+
+Create and include a new file named `custom_renderer.js`.
+
+Define a renderer that extends the base renderer.
+```js
+CustomRenderer = function(name) {
+  CustomRenderer.superClass_.constructor.call(this, name);
+};
+Blockly.utils.object.inherits(CustomRenderer,
+    Blockly.blockRendering.Renderer);
+```
+
+Register the renderer by name:
+```js
+Blockly.blockRendering.register('custom_renderer', CustomRenderer);
+```
+
+Add your renderer to the "Renderer" dropdown in `playground.html`.
+```html
+<select name="renderer" onchange="document.forms.options.submit()">
+  <option value="custom_renderer">Custom Renderer</option>
+  <option value="geras">Geras</option>
+  <option value="thrasos">Thrasos</option>
+  <option value="zelos">Zelos</option>
+  <option value="minimalist">Minimalist</option>
+</select>
+```
+
+To test, open the playground in a browser and select your new renderer from the dropdown.
+### The result
+
+The resulting block looks close to normal, but with slightly different alignment than Geras or Thrasos.
+
+![](assets/custom-renderer/custom_renderer.png)
+
+## Custom constants
+
+This sample shows the minumum code needed to define a new renderer that overrides constants from the base renderer.
+
+### Define your renderer
+Create and include a new file named `custom_constants.js`, and add it to the renderer dropdown.
+
+Define a renderer that extends the base renderer: `CustomConstantsRenderer`
```

Comment:
Just liked the wording from below a bit better and figured it could also fit up here: 
Define a renderer that extends the base renderer: `CustomConstantsRenderer ` and register it with the name `custom_constants `.